### PR TITLE
Don't merge boundary event subscriptions on instance migration

### LINKED ISSUE #19960: Ensure instance migration can't merge two boundary event subscriptions
Part of #19640 

>Map two message boundary events in the source to the same message boundary event in the target.

### PR DESCRIPTION
**Pull request by korthout**
## Description

When [migrating a process instance](https://docs.camunda.io/docs/components/concepts/process-instance-migration/), the user provides mapping instructions to clarify how the migration should be executed. This includes the ability to map boundary events. The subscription of an active element to a boundary event is migrated over if a mapping is provided for it (see [Dealing with attached boundary events](https://docs.camunda.io/docs/components/concepts/process-instance-migration/#dealing-with-attached-boundary-events)).

It should not be possible for a mapped element's boundary events to be merged into a single boundary event. This would mean an element instance is subscribed multiple times to the same boundary event. 

To avoid this, we check each boundary event attached to an active element and ensure that
they are the target of a mapping instruction only once.

This rejects process instance migration commands when a boundary event is the target of multiple mapping instructions.

## Related issues

closes #19960

### CODE REVIEW

**Comment by korthout** [comment]
No need to backport this to 8.5, because 8.5 does not support mapping message boundary events yet.

**Review by berkaycanbc — APPROVED** [review]
Thanks for the PR @korthout! 🚀 

I added two possible improvements but they are not blockers. Feel free to merge with or without addressing my suggestions.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceMigrationPreconditions.java [hunk 1]
```diff
@@ -615,6 +616,63 @@ public static void requireMappedBoundaryEventsToStayAttachedToSameElement(
...
+      final long processInstanceKey,
+      final DeployedProcess sourceProcessDefinition,
+      final String sourceElementId,
+      final Map<String, String> mappingInstructions) {
+    final var sourceElement =
+        sourceProcessDefinition
+            .getProcess()
+            .getElementById(sourceElementId, ExecutableCatchEventSupplier.class);
+
+    record Mapping(String sourceElementId, String targetElementId) {}
```
[the remark this answers is not in the document]
Comment (korthout) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceMigrationPreconditions.java:642]: Local records are implicitly static classes.

>A nested record class is implicitly static. That is, every member record class and local record class is static. ([source](https://docs.oracle.com/javase/specs/jls/se17/html/jls-8.html#jls-8.10))

I imagine that the JVM optimizes this away for us by loading it once and then simply using it. There should be no difference between a locally defined record and a regular record class.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceMigrationPreconditions.java [hunk 2]
```diff
@@ -615,6 +616,63 @@ public static void requireMappedBoundaryEventsToStayAttachedToSameElement(
...
+
+    final Map<String, List<String>> sourceBoundaryEventIdsByTargetBoundaryEventId =
```
[the remark this answers is not in the document]
Comment (korthout) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceMigrationPreconditions.java:644]: 👍 I really like this 🚀  Much easier to read

**Pull request merged**